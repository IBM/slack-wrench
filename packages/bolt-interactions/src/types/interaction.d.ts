import { App, SlackAction, SlackActionMiddlewareArgs } from '@slack/bolt';

import { InteractionFlow } from '../interaction-flow';

declare namespace Interaction {
  interface Controller<FlowState> {
    (flow: InteractionFlow<FlowState>, app: App): void;
  }

  type FlowIds = Record<string, string>;

  interface FlowContext<FlowState> {
    setState: (state: any, expiresAt?: number) => Promise<unknown>;
    state: FlowState;
    endFlow: () => Promise<unknown>;
    interactionIds: FlowIds;
  }

  interface FlowMiddlewareArgs<FlowState> {
    context: Record<string, string> & FlowContext<FlowState>;
  }

  interface FlowActionMiddlewareArgs<
    FlowState,
    ActionType extends SlackAction = SlackAction
  >
    extends FlowMiddlewareArgs<FlowState>,
      SlackActionMiddlewareArgs<ActionType> {}

  interface ActionConstraints {
    action_id: string;
    block_id?: string | RegExp;
    callback_id?: string | RegExp;
  }
}
