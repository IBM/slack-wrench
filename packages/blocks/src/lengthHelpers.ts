import R, {
  concat,
  curry,
  flip,
  has,
  ifElse,
  lensIndex,
  lensProp,
  lt,
  mapObjIndexed,
  over,
  pipe,
  prop,
  take,
  view,
} from 'ramda';

const dot3 = '...';

/**
 * given a truncate total limit and a string, returns a string with at most `limit` characters with '...' appended to the end
 * example:
 * ellipsisText('Hello World!', 8) => 'Hello...'
 */
const ellipsisText = curry((limit: number, value: string): string =>
  pipe<string, string, string>(
    take(limit - dot3.length),
    flip(concat)(dot3),
  )(value),
);

/**
 * takes string or { text: '...', ...}, and truncates either the string itself or the text property on the object
 * resulting string will be a maximum of `limit` characters.
 * the last 3 characters of the string will be `...` if the string is longer than `limit`
 */
export const ellipsis = <T>(limit: number, value: T): T =>
  ifElse(
    has('text'),
    over(lensProp('text'), ellipsisText(limit)),
    ellipsisText(limit),
  )(value);

/**
 * error function - throws error immediately on function call
 */
export const disallow = (limit: number, value: any): never => {
  throw Error(
    `Invalid length for property, max ${limit}: ${value.text || value}`,
  );
};

/**
 *  identity function which just swallows the `limit`, and won't truncate
 */
export const identity = <T>(limit: number, value: T): T => R.identity(value);

/**
 * truncates by just taking the first `limit` elements of `value`
 * works for list or string
 */
// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
export const truncate = <T>(limit: number, value: T): T => take(limit)(value);

const longerThan = curry((limit: number, value: any): boolean =>
  pipe<any, number, boolean>(prop('length'), lt(limit))(value),
);

/**
 * returns true if the relevant length on `value` is greater than `limit`
 */
const isTooLongForBlock = curry((limit: number, value: any): boolean =>
  ifElse(
    has('text'),
    pipe(prop('text'), longerThan(limit)),
    longerThan(limit),
  )(value),
);

export type TruncateFunction = <T>(limit: number, value: T) => T;

/**
 * mapping of properties to limit number and truncate function
 */
export type TruncateOptions = Record<string, [number, TruncateFunction]>;

/**
 * mostly internal function for handling extracting limits from TruncateOptions
 */
export const truncLimits: (
  options: TruncateOptions,
) => Record<string, number> = mapObjIndexed(view(lensIndex(0)));

/**
 * mostly internal function for handling extracting functions from TruncateOptions
 */
export const truncators: (
  options: TruncateOptions,
) => Record<string, TruncateFunction> = mapObjIndexed(view(lensIndex(1)));

/**
 * mostly internal function for building blocks - takes the element `obj` and
 * applies the truncation `functions` to each limited field iff that field's length is greater than its
 * respective limit for that block
 * @param obj
 * @param functions
 * @param limits
 */
export const applyTruncations = <T extends Record<string, any>>(
  obj: T,
  functions: Record<string, TruncateFunction>,
  limits: Record<string, number>,
): T => {
  const truncateKeys = Object.keys(functions);

  // for each key in object, apply the function in functions associated with that key if exists
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  return mapObjIndexed(<J>(value: J, key: string): J => {
    if (truncateKeys.includes(key) && isTooLongForBlock(limits[key])(value)) {
      return functions[key](limits[key], value);
    }

    return value;
  }, obj);
};
