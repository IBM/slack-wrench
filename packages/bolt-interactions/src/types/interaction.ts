import {
  App,
  SlackAction,
  SlackActionMiddlewareArgs,
  SlackViewAction,
  SlackViewMiddlewareArgs,
} from '@slack/bolt';

import { InteractionFlow } from '../interaction-flow';

export interface Controller<FlowState> {
  (flow: InteractionFlow<FlowState>, app: App): void;
}

export type FlowIds = Record<string, string>;

export interface FlowContext<FlowState> {
  setState: (state: any, expiresAt?: number) => Promise<unknown>;
  state: FlowState;
  endFlow: () => Promise<unknown>;
  interactionIds: FlowIds;
}

export interface FlowMiddlewareArgs<FlowState> {
  context: Record<string, string> & FlowContext<FlowState>;
}

export interface FlowActionMiddlewareArgs<
  FlowState,
  ActionType extends SlackAction = SlackAction
>
  extends FlowMiddlewareArgs<FlowState>,
    SlackActionMiddlewareArgs<ActionType> {}

export interface FlowViewMiddlewareArgs<
  FlowState,
  ViewActionType extends SlackViewAction = SlackViewAction
>
  extends FlowMiddlewareArgs<FlowState>,
    SlackViewMiddlewareArgs<ViewActionType> {}

export interface ActionConstraints {
  action_id: string;
  block_id?: string | RegExp;
  callback_id?: string | RegExp;
}
