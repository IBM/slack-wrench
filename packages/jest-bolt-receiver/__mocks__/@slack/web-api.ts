import mockWebApi from '@slack-wrench/jest-mock-web-client';

module.exports = mockWebApi(jest);
