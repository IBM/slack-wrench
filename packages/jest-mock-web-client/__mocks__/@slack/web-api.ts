import mockWebApi from '../../src/index';

module.exports = mockWebApi(jest);
