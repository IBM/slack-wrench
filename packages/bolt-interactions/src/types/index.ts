import * as Interaction from './interaction';

// eslint-disable-next-line import/prefer-default-export
export { Interaction };
