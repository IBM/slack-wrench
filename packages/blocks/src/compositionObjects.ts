import { MrkdwnElement, Option, PlainTextElement } from '@slack/types';
import { mergeLeft } from 'ramda';

import {
  applyTruncations,
  disallow,
  ellipsis,
  truncate,
  TruncateFunction,
  TruncateOptions,
  truncators,
  truncLimits,
} from './lengthHelpers';

// Composition Object Helpers --- https://api.slack.com/reference/block-kit/composition-objects

// --- Text Object ---  https://api.slack.com/reference/block-kit/composition-objects#text
export const Markdown = (text: string): MrkdwnElement => ({
  type: 'mrkdwn',
  text,
});

export const PlainText = (text: string, emoji = true): PlainTextElement => ({
  type: 'plain_text',
  text,
  emoji,
});

// --- Confirm Object --- https://api.slack.com/reference/block-kit/composition-objects#confirm

// --- Option Object --- https://api.slack.com/reference/block-kit/composition-objects#option
const optionTruncates: TruncateOptions = {
  text: [75, ellipsis],
  value: [75, disallow],
  description: [75, ellipsis],
  url: [3000, truncate],
};

export const OptionObject = (
  text: string,
  value: string,
  optionBlock: Partial<Option> = {},
  truncateFunctions: Record<string, TruncateFunction> = {},
): Option =>
  applyTruncations<Option>(
    {
      text: PlainText(text),
      value,
      ...optionBlock,
    } as Option,
    mergeLeft(truncateFunctions, truncators(optionTruncates)),
    truncLimits(optionTruncates),
  );

// --- Option Group Object --- https://api.slack.com/reference/block-kit/composition-objects#option_group

// --- Filter Object --- https://api.slack.com/reference/block-kit/composition-objects#filter_conversations
