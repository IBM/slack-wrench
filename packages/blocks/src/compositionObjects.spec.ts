import { PlainTextElement } from '@slack/types';

import { Markdown, OptionObject, PlainText } from './compositionObjects';
import { truncate } from './lengthHelpers';

const dynamicText = '0123456789'.repeat(301); // 3010 characters long
const dynamicTextElement: PlainTextElement = {
  type: 'plain_text',
  text: dynamicText,
};

describe('Element composition', () => {
  const text = 'Ravenclaw is obviously the best house.';

  describe('markdown object', () => {
    it('renders', () => {
      expect.assertions(1);
      expect(Markdown(text)).toMatchSnapshot();
    });
  });

  describe('plaintext object', () => {
    it('renders', () => {
      expect.assertions(1);
      expect(PlainText(text)).toMatchSnapshot();
    });

    it('renders without emoji', () => {
      expect.assertions(1);
      expect(PlainText(text, false)).toMatchSnapshot();
    });
  });

  describe('option object', () => {
    it('renders', () => {
      expect.assertions(1);
      expect(OptionObject('Why best?', 'why-best-raven')).toMatchSnapshot();
    });

    it('truncates ellipsis for text, description', () => {
      expect.assertions(2);
      const option = OptionObject(dynamicText, 'why-best-raven', {
        description: dynamicTextElement,
      });
      expect(option.text).toEqual(PlainText(`${dynamicText.substr(0, 72)}...`));
      expect(option.description).toEqual({
        ...dynamicTextElement,
        text: `${dynamicTextElement.text.substr(0, 72)}...`,
      });
    });

    it('truncates url', () => {
      expect.assertions(1);
      const option = OptionObject('text', 'value', { url: dynamicText });
      expect(option.url).toHaveLength(3000);
    });

    it('disallows too long value', () => {
      expect.assertions(1);
      expect(() => {
        OptionObject('Why best?', dynamicText);
      }).toThrow();
    });

    it('allows override truncateOptions for too long value', () => {
      expect.assertions(1);
      const option = OptionObject('Why best?', dynamicText, undefined, {
        value: truncate,
      });
      expect(option.value).toHaveLength(75);
    });
  });
});
