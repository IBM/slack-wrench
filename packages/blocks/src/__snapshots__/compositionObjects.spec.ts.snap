// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element composition markdown object renders 1`] = `
Object {
  "text": "Ravenclaw is obviously the best house.",
  "type": "mrkdwn",
}
`;

exports[`Element composition option object allows override truncateOptions for too long value 1`] = `
Object {
  "text": Object {
    "emoji": true,
    "text": "Why best?",
    "type": "plain_text",
  },
  "value": "012345678901234567890123456789012345678901234567890123456789012345678901234",
}
`;

exports[`Element composition option object renders 1`] = `
Object {
  "text": Object {
    "emoji": true,
    "text": "Why best?",
    "type": "plain_text",
  },
  "value": "why-best-raven",
}
`;

exports[`Element composition option object truncates ellipsis for text, description 1`] = `
Object {
  "description": Object {
    "text": "012345678901234567890123456789012345678901234567890123456789012345678901...",
    "type": "plain_text",
  },
  "text": Object {
    "emoji": true,
    "text": "012345678901234567890123456789012345678901234567890123456789012345678901...",
    "type": "plain_text",
  },
  "value": "why-best-raven",
}
`;

exports[`Element composition option object truncates url 1`] = `
Object {
  "text": Object {
    "emoji": true,
    "text": "Why best?",
    "type": "plain_text",
  },
  "url": "012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789",
  "value": "why-best-raven",
}
`;

exports[`Element composition plaintext object renders 1`] = `
Object {
  "emoji": true,
  "text": "Ravenclaw is obviously the best house.",
  "type": "plain_text",
}
`;

exports[`Element composition plaintext object renders without emoji 1`] = `
Object {
  "emoji": false,
  "text": "Ravenclaw is obviously the best house.",
  "type": "plain_text",
}
`;
